### PR TITLE
MAKE-862: implement wire protocol service

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/MAKE-862

* Implement service for responding to MongoDB wire protocol messages.
* Implement client for testing purposes.
* Fix various minor issues with makefile.

- [x] ~Waiting on [MAKE-994](https://jira.mongodb.org/browse/MAKE-994) and [MAKE-996](https://jira.mongodb.org/browse/MAKE-996) to be merged.~

### CODE REVIEW

File: wire/service.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	for name, handler := range map[string]mrpc.HandlerFunc{
+		// Required initialization commands
+		IsMasterCommand:                s.isMaster,
+		WhatsMyURICommand:              s.whatsMyURI,
+		BuildinfoCommand:               s.buildInfo,
+		BuildInfoCommand:               s.buildInfo,
+		GetLogCommand:                  s.getLog,
+		ReplSetGetStatusCommand:        s.replSetGetStatus,
+		GetFreeMonitoringStatusCommand: s.getFreeMonitoringStatus,
+		ListCollectionsCommand:         s.listCollections,
```
Comment (tychoish) [inline wire/service.go]: can we move these into mrpc as a "registry" of required commands?

File: wire/service.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	writeResponse(ctx, w, resp, BuildInfoCommand)
+}
+
+func (s *service) getLog(ctx context.Context, w io.Writer, msg mongowire.Message) {
+	resp, err := makeGetLogResponse([]string{}).Message()
+	if err != nil {
+		writeErrorResponse(ctx, w, errors.Wrap(err, "could not make response"), GetLogCommand)
+		return
+	}
+	writeResponse(ctx, w, resp, GetLogCommand)
```
Comment (tychoish) [inline wire/service.go]: as in gimlet, does it make sense to put the response writing commands into mrpc?

File: wire/manager.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		writeErrorResponse(ctx, w, errors.New("could not make response"), ManagerIDCommand)
+		return
+	}
+	writeResponse(ctx, w, resp, ManagerIDCommand)
+}
+
+// managerCreateProcess replies with a document of the form:
+//     {"ok": bool, "info": document}
+func (s *service) managerCreateProcess(ctx context.Context, w io.Writer, msg mongowire.Message) {
+	req, err := ExtractCreateProcessRequest(msg)
```
Comment (tychoish) [inline wire/manager.go]: why are extract message requests public?

File: wire/service.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		CompleteCommand:                s.processComplete,
+		WaitCommand:                    s.processWait,
+		SignalCommand:                  s.processSignal,
+		RegisterSignalTriggerIDCommand: s.processRegisterSignalTriggerID,
+		RespawnCommand:                 s.processRespawn,
+		TagCommand:                     s.processTag,
+		GetTagsCommand:                 s.processGetTags,
+		ResetTagsCommand:               s.processResetTags,
+	} {
+		if err := s.RegisterOperation(&mongowire.OpScope{
```
Comment (tychoish) [inline wire/service.go]: while I think we definitely need these kinds of handler funcs in this option, should we open a ticket for mrpc that gives us a different interface for registering handlers that's more like gimlet's (rest/v2)'s framework that would be easier to test?

File: wire/service.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	svc := &service{
+		Service: mrpc.NewService(host, port),
+		manager: m,
+	}
+	if err := svc.registerHandlers(); err != nil {
+		return nil, errors.Wrap(err, "error registering handlers")
+	}
+
+	cctx, ccancel := context.WithCancel(context.Background())
+	go func() {
```
Comment (tychoish) [inline wire/service.go]: recovery hanlder?

File: wire/util.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+func messageToResponse(msg mongowire.Message, out interface{}) error {
+	doc, err := responseMessageToDocument(msg)
+	if err != nil {
+		return errors.Wrap(err, "could not read response")
+	}
+	b, err := doc.MarshalBSON()
+	if err != nil {
+		return errors.Wrap(err, "could not convert document to BSON")
+	}
+	if err := bson.Unmarshal(b, out); err != nil {
```
Comment (tychoish) [inline wire/util.go]: why are we roundtripping this?

File: wire/util.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	"io"
+	"time"
+
+	"github.com/evergreen-ci/birch"
+	"github.com/evergreen-ci/mrpc/mongowire"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/jasper"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
```
Comment (tychoish) [inline wire/util.go]: can we avoid mgobson?

File: wire/service.go (hunk 4 above)
Comment (Kimchelly) [reply wire/service.go]: I can open a ticket but don't know what different interface you have in mind. As far as I'm aware, the gimlet method of adding routes is basically the same thing.

**Pull request closed without merging**